File: implementation/src/main/java/ro/anud/xml_xsd/implementation/model/ToOption/ToOption.java
package ro.anud.xml_xsd.implementation.model.ToOption;
import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.*;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import ro.anud.xml_xsd.implementation.util.RawNode;

import java.util.*;
import java.util.stream.Stream;
import ro.anud.xml_xsd.implementation.util.Subscription;
import java.util.function.Consumer;
import java.util.stream.Collectors;

import static ro.anud.xml_xsd.implementation.util.LocalLogger.logEnter;
import static ro.anud.xml_xsd.implementation.util.LocalLogger.logReturn;
import static ro.anud.xml_xsd.implementation.util.LocalLogger.log;
import static ro.anud.xml_xsd.implementation.util.LocalLogger.logReturnVoid;

  @EqualsAndHashCode
  @ToString
  @Builder
  @NoArgsConstructor
  @AllArgsConstructor
  @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
  public class ToOption implements  ro.anud.xml_xsd.implementation.util.LinkedNode {

    public static ToOption fromRawNode(RawNode rawNode) {
      logEnter();
      var instance = new ToOption();
      instance.setRawNode(rawNode);
      instance.deserialize(rawNode);
      return logReturn(instance);
    }
    public static ToOption fromRawNode(RawNode rawNode, ro.anud.xml_xsd.implementation.util.LinkedNode parent) {
      logEnter();
      var instance = fromRawNode(rawNode);
      instance.setParentNode(parent);
      return logReturn(instance);
    }
    public static Optional<ToOption> fromRawNode(Optional<RawNode> rawNode, ro.anud.xml_xsd.implementation.util.LinkedNode parent) {
        logEnter();
        return logReturn(rawNode.map(o -> ToOption.fromRawNode(o, parent)));
    }
    public static List<ToOption> fromRawNode(List<RawNode> rawNodeList, ro.anud.xml_xsd.implementation.util.LinkedNode parent) {
      logEnter();
      List<ToOption> returnList = Optional.ofNullable(rawNodeList)
          .orElse(List.of())
          .stream()
          .map(o -> ToOption.fromRawNode(o, parent))
          .collect(Collectors.toList());
      return logReturn(returnList);
    }

    //Attributes
    private String nodeRuleRef;
    private Integer distance;
    private Optional<Integer> maxDistance;
    private Integer adjacentDepthLimit;

    //Children elements
    private Optional<ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations> distanceToProgressMultiplier = Optional.empty();
    private Optional<ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations> personProgressProperty = Optional.empty();

    @ToString.Exclude()
    @EqualsAndHashCode.Exclude()
    @JsonIgnore
    @Getter
    @Setter
    @Builder.Default
    private RawNode rawNode = new RawNode();

    @Getter
    @ToString.Exclude()
    @EqualsAndHashCode.Exclude()
    @JsonIgnore
    @Builder.Default
    private Optional<ro.anud.xml_xsd.implementation.util.LinkedNode> parentNode = Optional.empty();

    @Builder.Default
    private List<Consumer<Set<Object>>> onChangeList = new ArrayList<>();

    public String nodeName() {
      return "to_option";
    }

    public void childChanged(Set<Object> set) {
      set.add(this);
      onChangeList.forEach(consumer -> consumer.accept(set));
      parentNode.ifPresent(linkedNode -> linkedNode.childChanged(set));
    }

    private void triggerOnChange() {
      childChanged(new HashSet<>());
    }

    public void setParentNode(ro.anud.xml_xsd.implementation.util.LinkedNode linkedNode) {
      this.parentNode = Optional.of(linkedNode);
      triggerOnChange();
    }

    public void removeChild(Object object) {
        if(object instanceof ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations) {
          this.distanceToProgressMultiplier = Optional.empty();
        }
        if(object instanceof ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations) {
          this.personProgressProperty = Optional.empty();
        }
    }

    public int buildIndexForChild(Object object) {
        if(object instanceof ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations) {
          return 0;
        }
        if(object instanceof ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations) {
          return 0;
        }
        return 0;
    }

    public void removeFromParent() {
      parentNode.ifPresent(node -> node.removeChild(this));
    }

    public Subscription onChange(Consumer<Set<Object>> onChange) {
      logEnter();
      onChangeList.add(onChange);
      return logReturn(() -> onChangeList.remove(onChange));
    }

    public void deserialize (RawNode rawNode) {
      var logger = logEnter();
      this.rawNode = rawNode;
      // Godot.GD.Print("Deserializing to_option");
      var innerLogger = logger.log("attributes");
      //Deserialize attributes
      innerLogger.log("node_rule_ref");
      this.nodeRuleRef = rawNode.getAttributeRequired("node_rule_ref");
      innerLogger.log("distance");
      this.distance = rawNode.getAttributeIntRequired("distance");
      innerLogger.log("maxDistance");
      this.maxDistance = rawNode.getAttributeInt("maxDistance");
      innerLogger.log("adjacent_depth_limit");
      this.adjacentDepthLimit = rawNode.getAttributeIntRequired("adjacent_depth_limit");
      innerLogger = logger.log("children");
      //Deserialize children
      innerLogger.log("distance_to_progress_multiplier");
      this.distanceToProgressMultiplier = ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations.fromRawNode(rawNode.getChildrenFirst("distance_to_progress_multiplier"), this);
      innerLogger.log("person_progress_property");
      this.personProgressProperty = ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations.fromRawNode(rawNode.getChildrenFirst("person_progress_property"), this);
      logReturnVoid();
    }

    public RawNode serializeIntoRawNode()
    {
      var logger = logEnter();
      var innerLogger = logger.log("attributes");
      //Serialize attributes
      innerLogger.log("node_rule_ref");
      rawNode.setAttribute("node_rule_ref", this.nodeRuleRef);
      innerLogger.log("distance");
      rawNode.setAttribute("distance", this.distance);
      innerLogger.log("maxDistance");
      this.maxDistance.ifPresent(o -> rawNode.setAttribute("maxDistance", o));
      innerLogger.log("adjacent_depth_limit");
      rawNode.setAttribute("adjacent_depth_limit", this.adjacentDepthLimit);

      innerLogger = logger.log("children");
      //Serialize children
      innerLogger.log("distance_to_progress_multiplier");
      rawNode.setChildren("distance_to_progress_multiplier", distanceToProgressMultiplier.stream().map(ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations::serializeIntoRawNode).toList());
      innerLogger.log("person_progress_property");
      rawNode.setChildren("person_progress_property", personProgressProperty.stream().map(ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations::serializeIntoRawNode).toList());
      return rawNode;
    }

    public void serialize(Document document, Element element)
    {
        // Godot.GD.Print("Serializing to_option");
        var updatedRawNode = serializeIntoRawNode();
        updatedRawNode.populateNode(document, element);
    }

    public String getNodeRuleRef()
    {
      return this.nodeRuleRef;
    }
    public ToOption setNodeRuleRef(String value)
    {
      this.nodeRuleRef = value;
      triggerOnChange();
      return this;
    }
    public Integer getDistance()
    {
      return this.distance;
    }
    public ToOption setDistance(Integer value)
    {
      this.distance = value;
      triggerOnChange();
      return this;
    }
    public Optional<Integer> getMaxDistance()
    {
      return this.maxDistance;
    }
    public ToOption setMaxDistance(Optional<Integer> value)
    {
      this.maxDistance = value;
      triggerOnChange();
      return this;
    }
    public Integer getAdjacentDepthLimit()
    {
      return this.adjacentDepthLimit;
    }
    public ToOption setAdjacentDepthLimit(Integer value)
    {
      this.adjacentDepthLimit = value;
      triggerOnChange();
      return this;
    }
    public Optional<ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations> getDistanceToProgressMultiplier()
    {
      return this.distanceToProgressMultiplier;
    }
    public ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations getDistanceToProgressMultiplierOrDefault()
    {
      return this.distanceToProgressMultiplier.orElseGet(() -> {
        var instance = new ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations();
        instance.setParentNode(this);
        this.distanceToProgressMultiplier = Optional.of(instance);
        return this.distanceToProgressMultiplier.get();
      });
    }
    public Stream<ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations> streamDistanceToProgressMultiplierOrDefault()
    {
      return Stream.of(getDistanceToProgressMultiplierOrDefault());
    }
    public Stream<ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations> streamDistanceToProgressMultiplier()
    {
      return distanceToProgressMultiplier.stream();
    }
    public ToOption setDistanceToProgressMultiplier(ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations value)
    {
      this.distanceToProgressMultiplier = Optional.ofNullable(value);
      value.setParentNode(this);
      triggerOnChange();
      return this;
    }

    public Optional<ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations> getPersonProgressProperty()
    {
      return this.personProgressProperty;
    }
    public ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations getPersonProgressPropertyOrDefault()
    {
      return this.personProgressProperty.orElseGet(() -> {
        var instance = new ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations();
        instance.setParentNode(this);
        this.personProgressProperty = Optional.of(instance);
        return this.personProgressProperty.get();
      });
    }
    public Stream<ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations> streamPersonProgressPropertyOrDefault()
    {
      return Stream.of(getPersonProgressPropertyOrDefault());
    }
    public Stream<ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations> streamPersonProgressProperty()
    {
      return personProgressProperty.stream();
    }
    public ToOption setPersonProgressProperty(ro.anud.xml_xsd.implementation.model.Type_mathOperations.Type_mathOperations value)
    {
      this.personProgressProperty = Optional.ofNullable(value);
      value.setParentNode(this);
      triggerOnChange();
      return this;
    }

  }


  /*
    dependant type:
    {
      "type": "element",
      "value": {
        "metaType": "object",
        "attributes": {
          "metaType": "object",
          "value": {
            "node_rule_ref": {
              "metaType": "primitive",
              "value": "xs:string",
              "isNullable": false
            },
            "distance": {
              "metaType": "primitive",
              "value": "xs:int",
              "isNullable": false
            },
            "maxDistance": {
              "metaType": "primitive",
              "value": "xs:int",
              "isNullable": true
            },
            "adjacent_depth_limit": {
              "metaType": "primitive",
              "value": "xs:int",
              "isNullable": false
            }
          }
        },
        "isSingle": false,
        "value": {
          "distance_to_progress_multiplier": {
            "metaType": "reference",
            "value": "type__math_operations",
            "isSingle": true,
            "isNullable": true
          },
          "person_progress_property": {
            "metaType": "reference",
            "value": "type__math_operations",
            "isSingle": true,
            "isNullable": true
          }
        },
        "isNullable": true
      },
      "name": "to_option"
    }
  */

File: gui-client/world_step/type__link_group/to_option/to_option.cs
using System;
using System.Collections.Immutable;
using System.Collections.Generic;
using System.Xml;
using System.Linq;
using Guiclient.util;
using Godot;
using XSD;

namespace XSD.Ntype__link_group.Nto_option {}
namespace XSD {
}
namespace XSD.Ntype__link_group {
  public class to_option : IEquatable<to_option>, XSD.ILinkedNode  {

    public static string ClassTypeId = ".type__link_group.to_option";
    public static string TagName = "to_option";

    public string NodeName {get =>"to_option";}
    public RawNode rawNode = new RawNode();

    private ILinkedNode? _parentNode;
    public ILinkedNode? ParentNode {get => _parentNode; set => _parentNode = value;}
    private List<Action<to_option>> _onSelfChangeCallbackList = new();
    private List<Action<List<ILinkedNode>>> _onChangeCallbackList = new();

    //Attributes
    private System.String _node_rule_ref;
    public System.String node_rule_ref { get => _node_rule_ref; set => _node_rule_ref = value; }
    private System.Int32 _distance;
    public System.Int32 distance { get => _distance; set => _distance = value; }
    private System.Int32? _maxDistance;
    public System.Int32? maxDistance { get => _maxDistance; set => _maxDistance = value; }
    private System.Int32 _adjacent_depth_limit;
    public System.Int32 adjacent_depth_limit { get => _adjacent_depth_limit; set => _adjacent_depth_limit = value; }

    //Children elements
    private type__math_operations? _distance_to_progress_multiplier = null;
    public type__math_operations distance_to_progress_multiplierOrCreate
    {
      get
      {
        if(_distance_to_progress_multiplier == null)
        {
          _distance_to_progress_multiplier = new();
          _distance_to_progress_multiplier.ParentNode = this;
          NotifyChange();
        }
        return _distance_to_progress_multiplier;
      }
      set
      {
        _distance_to_progress_multiplier = value;
        if(value != null)
        {
          value.ParentNode = this;
        }

      }
    }
    public type__math_operations? distance_to_progress_multiplier
    {
      get
      {
        return _distance_to_progress_multiplier;
      }
      set
      {
        _distance_to_progress_multiplier = value;
        if(value != null)
        {
          value.ParentNode = this;
        }
      }
    }
    public to_option()
    {
    }

    public to_option(RawNode rawNode)
    {
      Deserialize(rawNode);
    }

    public to_option(XmlElement xmlElement)
    {
      this.rawNode.Deserialize(xmlElement);
      Deserialize(rawNode);
    }

    public void SetAttribute(string name, string? value)
    {
      if(name == "node_rule_ref")
      {
        Set_node_rule_ref(value);
      }
      if(name == "distance")
      {
        Set_distance(value?.ToInt() ?? 0);
      }
      if(name == "maxDistance")
      {
        Set_maxDistance(value?.ToInt() ?? 0);
      }
      if(name == "adjacent_depth_limit")
      {
        Set_adjacent_depth_limit(value?.ToInt() ?? 0);
      }
    }

    public void SetChild(dynamic linkedNode)
    {
      if(linkedNode is type__math_operations distance_to_progress_multiplier)
      {
        this.distance_to_progress_multiplier = distance_to_progress_multiplier;
      }

    }

    public void ClearChild(dynamic linkedNode)
    {
      if(linkedNode is type__math_operations)
      {
        this.distance_to_progress_multiplier = null;
      }

    }

    public Action OnSelfChange(Action<to_option> callback)
    {
      _onSelfChangeCallbackList.Add(callback);
      return () => _onSelfChangeCallbackList.Remove(callback);
    }

    public Action OnSelfChangeNode(Action<ILinkedNode> callback)
    {
      _onSelfChangeCallbackList.Add(callback);
      return () => _onSelfChangeCallbackList.Remove(callback);
    }


    public Action OnChange(Action<List<ILinkedNode>> callback)
    {
      _onChangeCallbackList.Add(callback);
      return () => _onChangeCallbackList.Remove(callback);
    }

    public void Deserialize (RawNode rawNode)
    {
      this.rawNode = rawNode;
      // Godot.GD.Print("Deserializing to_option");
      //Deserialize arguments
      if(rawNode.attributes.ContainsKey("node_rule_ref"))
      {
        var attribute_node_rule_ref = rawNode.attributes["node_rule_ref"];
        this.node_rule_ref = rawNode.attributes["node_rule_ref"];
      }
      if(rawNode.attributes.ContainsKey("distance"))
      {
        var attribute_distance = rawNode.attributes["distance"];
        this.distance = attribute_distance.ToInt();
      }
      if(rawNode.attributes.ContainsKey("maxDistance"))
      {
        var attribute_maxDistance = rawNode.attributes["maxDistance"];
        this.maxDistance = attribute_maxDistance?.ToInt();
      }
      if(rawNode.attributes.ContainsKey("adjacent_depth_limit"))
      {
        var attribute_adjacent_depth_limit = rawNode.attributes["adjacent_depth_limit"];
        this.adjacent_depth_limit = attribute_adjacent_depth_limit.ToInt();
      }

      //Deserialize children
      distance_to_progress_multiplier = rawNode.InitializeWithRawNode("distance_to_progress_multiplier", distance_to_progress_multiplier);
      NotifyChange();
    }

    public RawNode SerializeIntoRawNode()
    {
      //Serialize arguments
      if(this._node_rule_ref != null)
      {
        rawNode.attributes["node_rule_ref"] = this._node_rule_ref.ToString();
      }
      if(this._distance != null)
      {
        rawNode.attributes["distance"] = this._distance.ToString();
      }
      if(this._maxDistance != null)
      {
        rawNode.attributes["maxDistance"] = this._maxDistance?.ToString();
      }
      if(this._adjacent_depth_limit != null)
      {
        rawNode.attributes["adjacent_depth_limit"] = this._adjacent_depth_limit.ToString();
      }

      //Serialize children
      if(distance_to_progress_multiplier != null) {
        rawNode.children["distance_to_progress_multiplier"] = new List<RawNode> { distance_to_progress_multiplier.SerializeIntoRawNode() };
      }
      return rawNode;
    }

    public void Serialize(XmlElement element)
    {
        // Godot.GD.Print("Serializing to_option");
        var updatedRawNode = SerializeIntoRawNode();
        updatedRawNode.Serialize(element);
    }
    public System.String Get_node_rule_ref()
    {
      return this.node_rule_ref;
    }
    public void Set_node_rule_ref(System.String value)
    {
      this.node_rule_ref = value;
      this.NotifyChange();
    }
    public System.Int32 Get_distance()
    {
      return this.distance;
    }
    public void Set_distance(System.Int32 value)
    {
      this.distance = value;
      this.NotifyChange();
    }
    public System.Int32? Get_maxDistance()
    {
      return this.maxDistance;
    }
    public void Set_maxDistance(System.Int32? value)
    {
      this.maxDistance = value;
      this.NotifyChange();
    }
    public System.Int32 Get_adjacent_depth_limit()
    {
      return this.adjacent_depth_limit;
    }
    public void Set_adjacent_depth_limit(System.Int32 value)
    {
      this.adjacent_depth_limit = value;
      this.NotifyChange();
    }


    public void DeserializeAtPath(string xpath, RawNode rawNode)
    {
      if(xpath.StartsWith("."))
      {
        xpath = xpath.Substring(1);
      }
      if(xpath.StartsWith(type__math_operations.TagName))
      {
        this.distance_to_progress_multiplier ??= new type__math_operations();
        var childXPath = xpath.Substring(type__math_operations.TagName.Length + 3);
        this.distance_to_progress_multiplier.DeserializeAtPath(childXPath, rawNode);
        return;
      }

      Deserialize(rawNode);
    }

    public void NotifyChange(List<ILinkedNode> linkedNodes)
    {
      if(_parentNode == null)
        return;
      linkedNodes.Add(this);
      _onSelfChangeCallbackList.ForEach(action => action(this));
      _onChangeCallbackList.ForEach(action => action(linkedNodes));
      _parentNode.NotifyChange(linkedNodes);
    }

    public void NotifyChange()
    {
      NotifyChange(new ());
    }

    public int? BuildIndexForChild(ILinkedNode linkedNode)
    {
      if(linkedNode is type__math_operations casted_distance_to_progress_multiplier) {
        return 0;
      }
      return null;
    }

    public bool IsValidChildType(ILinkedNode candidateChild) {
      return candidateChild is type__math_operations
      || false;
    }

    public bool Equals(to_option? obj)
    {
        if (obj == null || GetType() != obj.GetType())
            return false;

        var other = (to_option)obj;
        return Equals(node_rule_ref, other.node_rule_ref) && Equals(distance, other.distance) && Equals(maxDistance, other.maxDistance) && Equals(adjacent_depth_limit, other.adjacent_depth_limit) && Equals(distance_to_progress_multiplier, other.distance_to_progress_multiplier);
    }

    public override int GetHashCode()
    {
        var acc = 0;

        acc = HashCode.Combine(acc, node_rule_ref);
        acc = HashCode.Combine(acc, distance);
        acc = HashCode.Combine(acc, maxDistance);
        acc = HashCode.Combine(acc, adjacent_depth_limit);
        acc = HashCode.Combine(acc, distance_to_progress_multiplier);
        return acc;
    }
  }
}